_posts/2020-10-02-producer-tuning.md

```diff
@@ -0,0 +1,307 @@
+retries=2147483647
+```
+The `retries` property sets the number of retries when resending a failed message request. That impressive-looking number is the default and maximum value.
+
+> **Delivery timeout**
+>
+>If you are using `delivery.timeout.ms` in your producer configuration, producer requests will fail before the number of retries has been used if the timeout expires before a successful acknowledgment. The `delivery.timeout.ms` sets a limit on the time to wait for an acknowledgment of the success or failure to deliver a message. You can choose to leave `retries` unset and use `delivery.timeout.ms` to perform a similar function instead.
+
+There is a performance cost to introducing additional checks to the order of delivery.
+
+If you prefer not to use `acks=all` and idempotency,
+a lighter option is to set the number of in-flight requests to 1 (the default is 5) to preserve ordering.
+
+```properties
+# ...
+enable.idempotence=false
+max.in.flight.requests.per.connection=1
+retries=2147483647
+```
+
+In this way you avoid a situation where _Message-A_ fails only to succeed after _Message-B_ was already written to the broker.
+
+### Reliability guarantees
+
+Idempotence on its own is useful for exactly once writes to a single partition.
+
+But how do we guarantee the reliability of message delivery for exactly once writes for a set of messages across multiple partitions? We use idempotence again, but combine it with a unique transactional ID defined for the producer. Transactions guarantee that messages using the same transactional ID are produced once, and either _all_ are successfully written to the respective logs or _none_ of them are.
+
+You specify a unique transactional ID in the producer configuration,
+and also set the maximum allowed time for transactions in milliseconds before a timeout error is returned. The default is `900000` or 15 minutes.
+
+```properties
+# ...
+enable.idempotence=true
+max.in.flight.requests.per.connection=5
+acks=all
+retries=2147483647
+transactional.id=_UNIQUE-ID_
+transaction.timeout.ms=900000
+# ...
+```
+
+The transactional ID is registered with the Kafka cluster on the first operation, as well as a producer `epoch` checkpoint number, which is used to identify the active producer instance.
+
+Why might we want to identify the active producer? Say an application determines that a producer has failed and creates a new producer instance to restart a transaction. If both producers are now sending messages, duplicate records are being created and we have lost our exactly once integrity.
+
+By specifying a transaction ID, if a new producer instance starts, older instances of the producer are identified by their older `epoch` number and _fenced-off_ by Kafka so that their messages are no included. This maintains the integrity of the message passing by ensuring that there is only ever one valid producer with the transactional ID. Each `transactional.id` should be used for a unique set of topic partitions.
```

Comment:
```suggestion
By specifying a transaction ID, if a new producer instance starts, older instances of the producer are identified by their older `epoch` number and _fenced-off_ by Kafka so that their messages are not included. This maintains the integrity of the message passing by ensuring that there is only ever one valid producer with the transactional ID. Each `transactional.id` should be used for a unique set of topic partitions.
```